internal/service/ipam/api_subnet_data_source.go

```diff
@@ -0,0 +1,116 @@
+
+	bloxoneclient "github.com/infobloxopen/bloxone-go-client/client"
+	"github.com/infobloxopen/bloxone-go-client/ipam"
+	"github.com/infobloxopen/terraform-provider-bloxone/internal/flex"
+	"github.com/infobloxopen/terraform-provider-bloxone/internal/utils"
+)
+
+// Ensure provider defined types fully satisfy framework interfaces.
+var _ datasource.DataSource = &SubnetDataSource{}
+
+func NewSubnetDataSource() datasource.DataSource {
+	return &SubnetDataSource{}
+}
+
+// SubnetDataSource defines the data source implementation.
+type SubnetDataSource struct {
+	client *bloxoneclient.APIClient
+}
+
+func (d *SubnetDataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_" + "ipam_subnets"
+}
+
+type IpamsvcSubnetModelWithFilter struct {
+	Filters    types.Map  `tfsdk:"filters"`
+	TagFilters types.Map  `tfsdk:"tag_filters"`
+	Results    types.List `tfsdk:"results"`
+}
+
+func (m *IpamsvcSubnetModelWithFilter) FlattenResults(ctx context.Context, from []ipam.IpamsvcSubnet, diags *diag.Diagnostics) {
+	if len(from) == 0 {
+		return
+	}
+	m.Results = flex.FlattenFrameworkListNestedBlock(ctx, from, IpamsvcSubnetAttrTypes, diags, FlattenIpamsvcSubnet)
+}
+
+func (d *SubnetDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "",
+		Attributes: map[string]schema.Attribute{
+			"filters": schema.MapAttribute{
+				Description: "Filter are used to return a more specific list of results. Filters can be used to match resources by specific attributes, e.g. name. If you specify multiple filters, the results returned will have only resources that match all the specified filters.",
+				ElementType: types.StringType,
+				Optional:    true,
+			},
+			"tag_filters": schema.MapAttribute{
+				Description: "Tag Filters are used to filter by tags return a more specific list of results. If you specify multiple filters, the results returned will have only resources that match all the specified filters.",
+				ElementType: types.StringType,
+				Optional:    true,
+			},
+			"results": schema.ListNestedAttribute{
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: utils.DataSourceAttributeMap(IpamsvcSubnetResourceSchemaAttributes, &resp.Diagnostics),
+				},
+				Computed: true,
+			},
+		},
+	}
+}
+
+func (d *SubnetDataSource) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*bloxoneclient.APIClient)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected DataSource Configure Type",
+			fmt.Sprintf("Expected *bloxoneclient.APIClient, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
```

Comment:
Don't we repeat these lines in every datasource and same for resources too. We should probably move it to a common package for all to use

Reply:
Yes, I am sure there might be more to restructure. We can maybe focus on it together as a story, and do it once all the objects are done. Right now this is generated code, so it isn't much work anyway.

Reply:
Yes, thats the idea. I don't expect you to do it in this PR. Just pointing it out